File: code/os_prelude.py
import sys
import json
import os
import urllib.request

# --- Protocol setup ---
# Reserve real stdout (fd 1) for JSON protocol messages.
# Redirect print() to stderr so debug output does not corrupt the protocol.
_proto_out = open(1, 'w', buffering=1)
sys.stdout = sys.stderr

_final_result = None
_output_files = []

_CALLBACK_URL = os.environ.get('_SANDBOX_CALLBACK_URL', '')
_EXECUTION_ID = os.environ.get('_SANDBOX_EXECUTION_ID', '')


def call_tool(name, args=None):
    """Call an agent tool via the Oasis callback URL and return the result.

    Blocks until the tool returns. Raises RuntimeError on tool failure.

    Example:
        data = call_tool('web_search', {'query': 'latest pandas release'})
        content = call_tool('file_read', {'path': 'config.yaml'})
    """
    if not _CALLBACK_URL:
        raise RuntimeError("call_tool: no callback URL configured")

    payload = json.dumps({
        "execution_id": _EXECUTION_ID,
        "name": name,
        "args": args or {},
    }).encode('utf-8')
    req = urllib.request.Request(
        _CALLBACK_URL,
        data=payload,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=120) as resp:
            body = resp.read()
    except urllib.error.HTTPError as e:
        raise RuntimeError(
            f"Tool '{name}': HTTP {e.code}: "
            f"{e.read().decode('utf-8', errors='replace')}"
        )
    except Exception as e:
        raise RuntimeError(f"Tool '{name}': request failed: {e}")

    result = json.loads(body)
    if result.get("error"):
        raise RuntimeError(f"Tool '{name}' failed: {result['error']}")

    data = result.get("data", "")
    try:
        return json.loads(data) if isinstance(data, str) and data else data
    except (json.JSONDecodeError, TypeError):
        return data


def call_tools_parallel(calls):
    """Call multiple tools in parallel. Returns a list of results in order.

    Each element of calls is a (name, args) tuple.
    If any tool raises an exception, it is re-raised after all threads complete.

    Example:
        pages = call_tools_parallel([
            ('http_fetch', {'url': 'https://example.com/a'}),
            ('http_fetch', {'url': 'https://example.com/b'}),
        ])
    """
    import threading

    results = [None] * len(calls)
    errors = [None] * len(calls)

    def _worker(i, name, args):
        try:
            results[i] = call_tool(name, args)
        except Exception as exc:
            errors[i] = exc

    threads = [
        threading.Thread(target=_worker, args=(i, name, args or {}))
        for i, (name, args) in enumerate(calls)
    ]
    for t in threads:
        t.start()
    for t in threads:
        t.join()

    for err in errors:
        if err is not None:
            raise err

    return results


def set_result(data, files=None):
    """Set the structured result to return to the caller.

    Call this once at the end of your code. The data is JSON-serialized and
    returned in the 'output' field of the HTTP response.

    Args:
        data:  any JSON-serializable value (dict, list, str, int, etc.)
        files: optional list of relative file paths to include in the response
               as base64-encoded files. Example: ['chart.png', 'report.csv']

    Example:
        set_result({"summary": "done", "count": 42}, files=["chart.png"])
    """
    global _final_result, _output_files
    _final_result = data
    if files is not None:
        _output_files = list(files)


def install_package(name):
    """Install a Python package at runtime using pip.

    The container boundary provides isolation. Installed packages persist
    for the lifetime of the container.

    Example:
        install_package('httpx')
        import httpx
    """
    import subprocess as _sp
    print(f"[sandbox] pip install {name}...", file=sys.stderr)
    result = _sp.run(
        [sys.executable, '-m', 'pip', 'install', '-q', name],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        raise RuntimeError(f"pip install {name} failed: {result.stderr}")
    print(f"[sandbox] {name} installed", file=sys.stderr)


# --- User code starts below ---

File: code/os_prelude.js
// --- Protocol setup ---
// Reserve real stdout (fd 1) for JSON protocol messages.
// Redirect console.log/warn/error to stderr so debug output does not corrupt the protocol.
const _fs = require('fs');
const _path = require('path');
const _http = require('http');
const _https = require('https');

const _protoOut = _fs.createWriteStream('', { fd: 1 });
const _origConsole = { ...console };

// Redirect all console methods to stderr.
console.log = (...args) => process.stderr.write(args.map(String).join(' ') + '\n');
console.warn = (...args) => process.stderr.write(args.map(String).join(' ') + '\n');
console.error = (...args) => process.stderr.write(args.map(String).join(' ') + '\n');
console.info = (...args) => process.stderr.write(args.map(String).join(' ') + '\n');
console.debug = (...args) => process.stderr.write(args.map(String).join(' ') + '\n');

let _finalResult = undefined;
let _outputFiles = [];

const _CALLBACK_URL = process.env._SANDBOX_CALLBACK_URL || '';
const _EXECUTION_ID = process.env._SANDBOX_EXECUTION_ID || '';

/**
 * Call an agent tool via the Oasis callback URL and return the result.
 *
 * Blocks (async) until the tool returns. Throws on tool failure.
 *
 * @param {string} name - Tool name
 * @param {object} [args={}] - Tool arguments
 * @returns {Promise<any>} Parsed tool result
 *
 * @example
 *   const data = await callTool('web_search', { query: 'latest Node.js release' });
 *   const content = await callTool('file_read', { path: 'config.yaml' });
 */
function callTool(name, args) {
    if (!_CALLBACK_URL) {
        throw new Error('callTool: no callback URL configured');
    }

    const payload = JSON.stringify({
        execution_id: _EXECUTION_ID,
        name: name,
        args: args || {},
    });

    return new Promise((resolve, reject) => {
        const url = new URL(_CALLBACK_URL);
        const mod = url.protocol === 'https:' ? _https : _http;
        const req = mod.request(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Content-Length': Buffer.byteLength(payload),
            },
            timeout: 120000,
        }, (res) => {
            let body = '';
            res.on('data', (chunk) => { body += chunk; });
            res.on('end', () => {
                try {
                    const result = JSON.parse(body);
                    if (result.error) {
                        reject(new Error(`Tool '${name}' failed: ${result.error}`));
                        return;
                    }
                    let data = result.data || '';
                    if (typeof data === 'string' && data) {
                        try { data = JSON.parse(data); } catch (_) { /* use as-is */ }
                    }
                    resolve(data);
                } catch (e) {
                    reject(new Error(`Tool '${name}': invalid response: ${e.message}`));
                }
            });
        });
        req.on('error', (e) => reject(new Error(`Tool '${name}': request failed: ${e.message}`)));
        req.on('timeout', () => {
            req.destroy();
            reject(new Error(`Tool '${name}': request timed out`));
        });
        req.write(payload);
        req.end();
    });
}

/**
 * Call multiple tools in parallel. Returns results in the same order as input.
 *
 * @param {Array<[string, object]>} calls - Array of [name, args] tuples
 * @returns {Promise<any[]>} Array of results in order
 *
 * @example
 *   const [a, b] = await callToolsParallel([
 *       ['http_fetch', { url: 'https://example.com/a' }],
 *       ['http_fetch', { url: 'https://example.com/b' }],
 *   ]);
 */
function callToolsParallel(calls) {
    return Promise.all(calls.map(([name, args]) => callTool(name, args || {})));
}

/**
 * Set the structured result to return to the caller.
 *
 * Call this once at the end of your code. The data is JSON-serialized and
 * returned in the 'output' field of the HTTP response.
 *
 * @param {any} data - Any JSON-serializable value
 * @param {string[]} [files] - Optional file paths to include in response
 *
 * @example
 *   setResult({ summary: 'done', count: 42 }, ['chart.png']);
 */
function setResult(data, files) {
    _finalResult = data;
    if (files !== undefined) {
        _outputFiles = Array.isArray(files) ? files : [files];
    }
}

/**
 * Install an npm package at runtime.
 *
 * The container boundary provides isolation. Installed packages persist
 * for the lifetime of the container.
 *
 * @param {string} name - Package name (e.g. 'lodash', 'cheerio')
 *
 * @example
 *   await installPackage('cheerio');
 *   const cheerio = require('cheerio');
 */
function installPackage(name) {
    const { execSync } = require('child_process');
    console.log(`[sandbox] npm install ${name}...`);
    try {
        execSync(`npm install --no-save ${name}`, {
            stdio: ['pipe', 'pipe', 'pipe'],
            cwd: process.env._SANDBOX_WORKSPACE || process.cwd(),
        });
        console.log(`[sandbox] ${name} installed`);
    } catch (e) {
        throw new Error(`npm install ${name} failed: ${e.stderr ? e.stderr.toString() : e.message}`);
    }
}

// --- User code starts below ---
// The user code is wrapped in an async IIFE so top-level await works.
(async () => {
try {
